File: 0279/sol2.cpp

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int numSquares(int n) {
        while(n % 4 == 0) n /= 4;
        if (n % 8 == 7) return 4;
        for (int r1 = 0; r1 * r1 <= n; r1++) {
            int r2 = sqrt(n - r1 * r1);
            if (r1 * r1 + r2 * r2 == n) return !!r1 + !!r2;
        }

        return 3;
    }
};

int main() {
    int n = 12;
    Solution sol;

    cout << sol.numSquares(n) << endl;

    return 0;
}

File: 1492/sol2.py
import math
class Solution(object):
    def kthFactor(self, n, k):
        """
        :type n: int
        :type k: int
        :rtype: int
        """
        
        i = 1
        total = 0
        while i < math.sqrt(n):
            if n % i == 0:
                total += 1
            i += 1
        total *= 2
        if math.sqrt(n) == i:
            total += 1
            if k == total // 2 + 1:
                return i
            
        i = 1
        j = 0
        while i < math.sqrt(n):
            if n % i == 0:
                j += 1
                if j == k:
                    return i
                elif total + 1 - j == k:
                    return int(n / i)
            i += 1

        return -1

    
sol = Solution()

# test cases
n = 12
k = 3
# n = 7
# k = 2
# n = 4
# k = 4

print(sol.kthFactor(n, k))
